### ISSUE TITLE
Implicit interface conversion

### ISSUE THREAD

**Issue by Opiumtm** [dotnet/roslyn#14186]
Today's implementation of implicit conversion operators doesn't allow interfaces to participate - only abstract classes, which isn't always possible or desirable (as interface is generally better than abstract class if you're planning your type hierarchy).

Why it's important? It allow to use adapter pattern transparently. You can write or generate (using upcoming "source generators" feature) adapter for certain interface and if implicit interface conversion is possible you could simply omit explicit adapter instantiation and use classes (which have adapter for certain interface + implicit conversion operator to certain interface) just in place of adapted interface.

It's all possible if you use abstract classes (implicit conversion operators are only applicable for classes and structs), but not for interfaces. It's a huge miss as many people (including me) prefer to use interfaces instead of abstract classes to handle polymorphism when possible. 

**How "extension everything" upcoming language feature would handle implicit interface conversion case?**

**Comment by DavidArno** [comment]
I don't think the "extension everything" feature will, but a feature that is part of pattern matching (but which isn't being released with C# 7) will address this. A new operator, `is` was/will be added to the language such that you could define:

````cs
public static bool operator is(AdapterType adapter, out AdaptedType original)
{
    original = convert from adapter to original;
    return true if conversion works;
}
````

And it can then be used like:
````cs
if (original is AdapterType adapter)
{
    // user adapter here
}
````

**Comment by HaloFour** [comment]
That prohibition is intentional in order to avoid the ambiguity in attempting to resolve an assignment between a conversion and a simple cast.  The conversion operators are intended to bridge between two types which are otherwise incompatible with one another.  The spec further explicitly prohibits redefining existing conversion operations, which would include downcasts.

**Comment by gulbanana** [comment]
would it be safe to support only implicit conversion *from* interfaces?

**Comment by whoisj** [comment]
I just hit this issue, and the restriction is making life rather difficult.

We've decided that, where possible, the library limits public types to interfaces. The unfortunate part is that classes cannot implement `public static explicit operator IThing(Thing value)` because of "reasons". I understand the difficulty of deciding which cast operator to use, but not allowing explicit casts is ridiculous.

Here's the actual scenario. We have two interfaces `ITypeA` and `ITypeB`, and a class which implements both `class ClassAB : ITypeA, ITypeB`. References to objects of `ClassAB` are given back as references to `ITypeA`. However, despite the reference is really `typeof(ClassAB)` any attempted to `itypeBref as ITypeB` always results in `null`. Makes very little sense.

**Comment by svick** [comment]
@whoisj What you're describing [should work](https://dotnetfiddle.net/WRm4r3). Are you sure there isn't something else going on in your code?

**Comment by whoisj** [comment]
> Are you sure there isn't something else going on in your code?

No. This is a large code base, there could be other things happening. However, it seems consistent to me. Perhaps there's some hierarchy of interface inheritance that is causing issues?

**Comment by HaloFour** [comment]
@whoisj 

That's not the behavior of C#.  The type of the instance doesn't change depending on the type of the variable.  Sounds like that project is wrapping those instances, either directly or via proxy, so that the instance you are working with are actually not instances of `ClassAB`.

**Comment by theunrepentantgeek** [comment]
I've run into this as well - from a different angle.

I'd defined an interface `IOption<T>` with two implementations, `Some<T>` and `None<T>`, and wanted to define a implicit cast to "wrap" a `T` into a `Some<T>`.

Essentially, I had this code:

    public IOption<Customer> FindCustomer()
    {
        Customer c = _database.FindCustomer();
        return new Some<Customer>(c);
    }

and I wanted to be able to write this:

    public IOption<Customer> FindCustomer()
    {
        Customer c = _database.FindCustomer();
        return c;
    }

which requires both defining an implicit cast from `T` to `Some<T>` and having the compiler recognize that this would be satisfactory.

**Comment by whoisj** [comment]
@theunrepentantgeek that is basically identical to what I was attempting to express, but significantly more clear and less confusing. Thanks 😄

**Comment by hrumhurum** [comment]
Just hit this unconvenient issue:

```c#
 struct Optional<TValue>
    {
        public Optional(TValue value)
        {
            _Value = value;
            _HasValue = true;
        }

        TValue _Value;
        bool _HasValue;

        public TValue Value
        {
            get
            {
                if (!_HasValue)
                    throw new Exception("Optional value not present.");
                return _Value;
            }
        }

        public bool HasValue => _HasValue;

        public static implicit operator Optional<TValue>(TValue value) => new Optional<TValue>(value);
    }
```

The usage scenarios:

```c#
Optional<int> a = 0; // works

Optional<string> b = null; // works

System.IO.Stream stream = null;
Optional<System.IO.Stream> c = stream; // works

IDisposable disposable = null;
Optional<IDisposable> c = disposable; // DOES NOT WORK
```
Is there any ongoing work to solve this in future C# versions?

**Comment by adnanalbeda** [comment]
I came across this problem today, three years after beginning this issue here.
Well, Microsoft has lunched preview version that makes interfaces have default implementation or values, I am not sure whether this problem would remain after publishing this feature.

**Comment by petriashev** [comment]
This feature can reduce annoying code. Is there any plans to implement it?

**Comment by biiiipy** [comment]
I'm also interesting in this. This seems like an arbitrary limitation.

**Comment by Jehoel** [comment]
I'd prefer it if C# supported implicit-conversion _and_ explicit-conversion through an interface that's part of the BCL, instead of through only the `static implicit operator` approach, that way we could use conversions in generic code (as generic constraints can use interfaces, but cannot require static members).

What I'd like to see is something like this as a standard interface:

```
namespace System
{
    // For representing domain identity, even when object-reference-identity cannot be maintained due to multiple-inheritance being disallowed.
    public interface IAm<T>
    {
        T Self { get; }
    }

    // For conversion that will always succeed.
    public interface IConvertTo<T>
    {
       T Convert();
    }

    // For conversions that might possibly fail.
    public interface IMaybeConvertTo<T>
    {
        Boolean TryConvertTo( [NotNullWhen(true)] out T converted );
    }
}
```

This would be useful in many cases and conveniently [side-steps Eric Lippert's discussion](https://stackoverflow.com/a/9231325/159145) about not wanting user-defined conversions to replace built-in conversions that preserve object-identity.

But this could be made _even better_ by adding a new C# keyword `is-maybe` which is aware of these interfaces and so performs the exhaustive `QueryService`-esque check for us:

```
if( something is-maybe AnotherType another )
{
    DoStuff( another );
}
```

Where the `is-maybe` operator is equivalent to:

```
{
    if( something is AnotherType a )
    {
        DoStuff( a );
    }
    else if( something is IAm<AnotherType> b )
    {
        DoStuff( b.Self );
    }
    else if( something is IConvertTo<AnotherType> c )
    {
        DoStuff( c.Convert() );
    }
    else if( something is IMaybeConvertTo<AnotherType> d && d.TryConvertTo( out AnotherType e  ) )
    {
        DoStuff( e );
    }
}
```

Right now I implement this `is-maybe` operator as a generic extension method (and define those interfaces in my own private utility library), but because we can't have custom control-structures (i.e. force-inlined lambdas and non-local returns) it's just as difficult and ugly to use as the inadequate code it's meant to replace...

**Comment by HaloFour** [comment]
@Jehoel 

> I'd prefer it if C# supported implicit-conversion and explicit-conversion through an interface that's part of the BCL, instead of through only the `static implicit operator` approach, that way we could use conversions in generic code (as generic constraints can use interfaces, but cannot require static members) as well as reduce boilerplate type conversion code.

The runtime and C# teams are currently exploring the possibility of [interfaces having static members](https://github.com/dotnet/csharplang/issues/4436), one of the primary use cases being to support operators for generic math, but it also would apply to generic conversion operators.  This is a part of a larger project called ["shapes"/"roles"](https://github.com/dotnet/csharplang/issues/1711) which expands on the use of interfaces as generic constraints even where the generic type argument doesn't implicit the interface directly.

> But this could be made even better by adding a new C# keyword `is-maybe` which is aware of these interfaces and so performs the exhaustive QueryService-esque check for us:

The `is` operator is already conditional, and there are proposals to add custom patterns to the language that would allow one type to match the pattern of another type: https://github.com/dotnet/csharplang/issues/1047

I think that would cover the cases you described.
